jupyterhub_carina/CarinaSpawner.py

```diff
@@ -0,0 +1,218 @@
+import docker
+from dockerspawner import DockerSpawner
+import os.path
+import re
+import requests
+import shutil
+from tornado import gen
+from traitlets import Dict, Integer, Unicode
+from .CarinaOAuthClient import CarinaOAuthClient
+
+
+class CarinaSpawner(DockerSpawner):
+    """
+    Spawn the user's Jupyter server on their Carina cluster.
+    """
+
+    cluster_name = Unicode(
+        'jupyterhub',
+        help="The name of the user's Carina cluster.",
+        config=True)
+
+    # Override the default container name. Since we are running on the user's cluster, we don't need the username suffix
+    container_name = Unicode(
+        'jupyter',
+        help="The name of the Jupyter server container running on the user's cluster.",
+        config=True)
+
+    cluster_polling_interval = Integer(
+        30,
```

Comment:
30 is good. For reference, Carina's "average" time for a build is about one minute (was 45 seconds before the introduction of consul). In `carina`, the httpTimeout is 15 seconds.